tree-sitter/src/TreeSitter/Unmarshal/Examples.hs

```diff
@@ -1,62 +1,93 @@
 import qualified Data.Text as Text
-import GHC.Generics ((:+:), Generic)
+import qualified Data.Text.Encoding as Text
+import GHC.Generics ((:+:), Generic1)
+import Numeric (readDec)
+import Prelude hiding (fail)
+import Source.Range
 import TreeSitter.Unmarshal
 
 -- | An example of a sum-of-products datatype.
 data Expr a
   = IfExpr (If a)
   | BlockExpr (Block a)
+  | VarExpr (Var a)
   | LitExpr (Lit a)
   | BinExpr (Bin a)
-  deriving (Generic, Unmarshal)
+  deriving (Generic1, Unmarshal)
 
 -- | Product with multiple fields.
 data If a = If { ann :: a, condition :: Expr a, consequence :: Expr a, alternative :: Maybe (Expr a) }
-  deriving (Generic, Unmarshal)
+  deriving (Generic1, Unmarshal)
 
-instance SymbolMatching (If a) where
+instance SymbolMatching If where
   symbolMatch _ _ = False
   showFailure _ _ = ""
 
 -- | Single-field product.
 data Block a = Block { ann :: a, body :: [Expr a] }
-  deriving (Generic, Unmarshal)
+  deriving (Generic1, Unmarshal)
 
-instance SymbolMatching (Block a) where
+instance SymbolMatching Block where
   symbolMatch _ _ = False
   showFailure _ _ = ""
 
 -- | Leaf node.
-data Lit a = Lit { ann :: a, text :: Text.Text }
-  deriving (Generic, Unmarshal)
+data Var a = Var { ann :: a, text :: Text.Text }
+  deriving (Generic1, Unmarshal)
 
-instance SymbolMatching (Lit a) where
+instance SymbolMatching Var where
+  symbolMatch _ _ = False
+  showFailure _ _ = ""
+
+-- | Custom leaf node.
+data Lit a = Lit { ann :: a, lit :: IntegerLit }
+  deriving (Generic1, Unmarshal)
+
+instance SymbolMatching Lit where
   symbolMatch _ _ = False
   showFailure _ _ = ""
 
 -- | Product with anonymous sum field.
 data Bin a = Bin { ann :: a, lhs :: Expr a, op :: (AnonPlus :+: AnonTimes) a, rhs :: Expr a }
-  deriving (Generic, Unmarshal)
+  deriving (Generic1, Unmarshal)
 
-instance SymbolMatching (Bin a) where
+instance SymbolMatching Bin where
   symbolMatch _ _ = False
   showFailure _ _ = ""
 
 -- | Anonymous leaf node.
 newtype AnonPlus a = AnonPlus { ann :: a }
-  deriving (Generic, Unmarshal)
+  deriving (Generic1, Unmarshal)
 
-instance SymbolMatching (AnonPlus a) where
+instance SymbolMatching AnonPlus where
   symbolMatch _ _ = False
   showFailure _ _ = ""
 
 -- | Anonymous leaf node.
 newtype AnonTimes a = AnonTimes { ann :: a }
-  deriving (Generic, Unmarshal)
+  deriving (Generic1, Unmarshal)
 
-instance SymbolMatching (AnonTimes a) where
+instance SymbolMatching AnonTimes where
   symbolMatch _ _ = False
   showFailure _ _ = ""
+
+
+newtype IntegerLit = IntegerLit Integer
+
+instance UnmarshalAnn IntegerLit where
+  unmarshalAnn node = do
+    Range start end <- unmarshalAnn node
+    bytestring <- ask
+    let drop = B.drop start
+        take = B.take (end - start)
+        slice = take . drop
+        str = Text.unpack (Text.decodeUtf8 (slice bytestring))
+    case readDec str of
+      (i, _):_ -> pure (IntegerLit i)
+      _        -> fail ("could not parse '" <> str <> "'")
```

Comment:
So, we tell it how: get the range, slice the source down to a string, and use `readDec` to produce an `Integer` _with total error-checking of malformed inputs_.

We’re unlikely to ever have the TH code produce ASTs with `Integer` fields—parsing the contents of these strings is language-specific, after all—but we could imagine having the TH code allow us to specify some datatypes that we want to define by hand instead. This gives us a way to do that which still usefully integrates with `Unmarshal`, meaning that you can derive _most_ of your data, specialize only small parts of it, only do this parsing once, right up front, and still parse and unmarshal the whole thing to Haskell data.